src/Analyzer.JsonRuleEngine/JsonRuleEngine.cs

```diff
@@ -40,26 +41,56 @@ public IEnumerable<IEvaluation> EvaluateRules(TemplateContext templateContext, s
 
             foreach(RuleDefinition rule in rules)
             {
-                var ruleExpression = rule.Evaluation.ToExpression();
-                var evaluation = ruleExpression.Evaluate(
+                var ruleExpression = rule.ExpressionDefinition.ToExpression();
+                JsonRuleEvaluation evaluation = ruleExpression.Evaluate(
                     new JsonPathResolver(
                         templateContext.ExpandedTemplate,
                         templateContext.ExpandedTemplate.Path));
 
+                 evaluation.RuleDefinition = rule;
+                 evaluation.FileIdentifier = templateContext.TemplateIdentifier;
+
                 // If there are no matching cases of the rule, do not create an evaluation
-                if (!(evaluation as IEvaluation).HasResults())
+                if (rule.ExpressionDefinition is LeafExpressionDefinition)
```

Comment:
Instead of looking at what kind of expression was evaluated, can you make HasResults recursive?  i.e. HasResults could look at the Results collection, and if that's empty, will see if any of its Evaluations have results.  That way the rule engine doesn't have to understand the details of how rules are evaluated, and I think will be less code overall.